scipy/sparse/csgraph/_shortest_path.pyx

```diff
@@ -655,47 +841,98 @@ cdef _dijkstra_undirected(
-                            if return_pred:
-                                pred[i, j_current] = v.index
-                        elif current_node.val > next_val:
-                            decrease_val(&heap, current_node, next_val)
-                            if return_pred:
-                                pred[i, j_current] = v.index
+            _dijkstra_scan_heap(&heap,
+                                v,
+                                nodes,
+                                csr_weights,
+                                csr_indices,
+                                csr_indptr,
+                                pred,
+                                return_pred,
+                                limit,
+                                i)
+
+            _dijkstra_scan_heap(&heap,
+                                v,
+                                nodes,
+                                csrT_weights,
+                                csrT_indices,
+                                csrT_indptr,
+                                pred,
+                                return_pred,
+                                limit,
+                                i)
 
             #v has now been scanned: add the distance to the results
             dist_matrix[i, v.index] = v.val
 
     free(nodes)
 
 
+                        
+
+cdef _dijkstra_undirected_multi(
+            int[:] source_indices,
+            double[:] csr_weights,
+            int[:] csr_indices,
+            int[:] csr_indptr,
+            double[:] csrT_weights,
+            int[:] csrT_indices,
+            int[:] csrT_indptr,
+            double[:] dist_matrix,
+            int[:] pred,
+            int[:] sources,
+            DTYPE_t limit):
+    cdef:
+        unsigned int Nind = source_indices.shape[0]
+        unsigned int N = dist_matrix.shape[0]
+        unsigned int i, k, j_source, j_current
+        ITYPE_t j
+        DTYPE_t next_val
+        int return_pred = (pred.size > 0)
+        FibonacciHeap heap
+        FibonacciNode *v
+        FibonacciNode *current_node
+        FibonacciNode* nodes = <FibonacciNode*> malloc(N *
+                                                        sizeof(FibonacciNode))
+
+    _dijkstra_setup_heap_multi(&heap,
+                               nodes,
+                               source_indices,
+                               sources,
+                               dist_matrix)
+    
+    while heap.min_node:
+        v = remove_min(&heap)
+        v.state = SCANNED
+
+        _dijkstra_scan_heap_multi(&heap,
+                                  v,
+                                  nodes,
+                                  csr_weights,
+                                  csr_indices,
+                                  csr_indptr,
+                                  pred,
+                                  sources,
+                                  return_pred,
+                                  limit)
```

Comment:
I think this long-list format actually hurts readability a bit, as it makes the file much longer than necessary. WDYT about doing:
```
        _dijkstra_scan_heap_multi(
            &heap, v, nodes, csr_weights, csr_indices, csr_indptr,
            pred, sources, return_pred, limit)
```

Reply:
agreed.. i tried to consolidate related variables on the same line, slightly different than your suggestion but i hope more readable.